pkg/controller/pxc/users.go

```diff
@@ -354,54 +355,69 @@ func (r *ReconcilePerconaXtraDBCluster) manageSysUsers(cr *api.PerconaXtraDBClus
+	if cr.Spec.ProxySQL == nil || !cr.Spec.ProxySQL.Enabled || cr.Status.PXC.Ready < 1 {
 		return nil
 	}
-	// sync users if ProxySql enabled
-	if cr.Spec.ProxySQL == nil || !cr.Spec.ProxySQL.Enabled || cr.Status.ObservedGeneration != cr.Generation || cr.Status.PXC.Ready < 1 {
+	if cr.Status.Status != api.AppStateReady || cr.Status.ProxySQL.Status != api.AppStateReady {
 		return nil
 	}
-	pod := corev1.Pod{}
-	err := r.client.Get(context.TODO(),
-		types.NamespacedName{
-			Namespace: cr.Namespace,
-			Name:      cr.Name + "-proxysql-0",
-		},
-		&pod,
-	)
-	if err != nil && k8serrors.IsNotFound(err) {
-		return err
-	} else if err != nil {
-		return errors.Wrap(err, "get proxysql pod")
-	}
-	var errb, outb bytes.Buffer
-	err = r.clientcmd.Exec(&pod, "proxysql", []string{"proxysql-admin", "--syncusers"}, nil, &outb, &errb, false)
-	if err != nil {
-		return errors.Errorf("exec syncusers: %v / %s / %s", err, outb.String(), errb.String())
-	}
-	if len(errb.Bytes()) > 0 {
-		return errors.New("syncusers: " + errb.String())
+
+	for i := 0; i < int(cr.Spec.PXC.Size); i++ {
+		pod := corev1.Pod{}
+		err := r.client.Get(context.TODO(),
+			types.NamespacedName{
+				Namespace: cr.Namespace,
+				Name:      cr.Name + "-proxysql-" + strconv.Itoa(i),
+			},
+			&pod,
+		)
+		if err != nil && k8serrors.IsNotFound(err) {
+			return err
+		} else if err != nil {
+			return errors.Wrap(err, "get proxysql pod")
+		}
+		var errb, outb bytes.Buffer
+		err = r.clientcmd.Exec(&pod, "proxysql", []string{"proxysql-admin", "--syncusers"}, nil, &outb, &errb, false)
+		if err != nil {
+			return errors.Errorf("exec syncusers: %v / %s / %s", err, outb.String(), errb.String())
+		}
+		if len(errb.Bytes()) > 0 {
+			return errors.New("syncusers: " + errb.String())
+		}
 	}
 
 	return nil
 }
 
-func updateProxyUsers(proxyUsers []users.SysUser, internalSysSecretObj *corev1.Secret, cr *api.PerconaXtraDBCluster) error {
+func (r *ReconcilePerconaXtraDBCluster) updateProxyUsers(proxyUsers []users.SysUser, internalSysSecretObj *corev1.Secret, cr *api.PerconaXtraDBCluster) error {
 	if len(proxyUsers) == 0 {
 		return nil
 	}
+	for i := 0; i < int(cr.Spec.ProxySQL.Size); i++ {
+		pod := corev1.Pod{}
+		err := r.client.Get(context.TODO(),
+			types.NamespacedName{
+				Namespace: cr.Namespace,
+				Name:      cr.Name + "-proxysql-" + strconv.Itoa(i),
+			},
+			&pod,
+		)
+		if err != nil && k8serrors.IsNotFound(err) {
+			return err
+		} else if err != nil {
+			return errors.Wrap(err, "get proxysql pod")
+		}
 
-	um, err := users.NewManager(cr.Name+"-proxysql-unready."+cr.Namespace+":6032", "proxyadmin", string(internalSysSecretObj.Data["proxyadmin"]))
-	if err != nil {
-		return errors.Wrap(err, "new users manager")
-	}
-	defer um.Close()
-
-	err = um.UpdateProxyUsers(proxyUsers)
-	if err != nil {
-		return errors.Wrap(err, "update proxy users")
+		um, err := users.NewManager(pod.Status.PodIP+":6032", "proxyadmin", string(internalSysSecretObj.Data["proxyadmin"]))
```

Comment:
please lets try in the next way and without fetching Pod
```suggestion
		um, err := users.NewManager(cr.Name + "-proxysql-" + strconv.Itoa(i)
+"."+cr.Name+"-proxysql-unready."+cr.Namespace+":6032", "proxyadmin", string(internalSysSecretObj.Data["proxyadmin"]))
```

Reply:
https://github.com/percona/percona-xtradb-cluster-operator/pull/795/commits/73ab8492c1953ae3cd796e194383109390277233